lib/yeti_exception/error.rb

```diff
@@ -0,0 +1,60 @@
+require 'json'
+
+module YetiException
+
+  # Enhanced exception class, with some features added beyond those in
+  # StandardError:
+  #
+  # - The class in which the exception was raised.  This can be useful for
+  #   logging, since the backtrace is much more verbose.
+  #
+  # - A details hash, containing arbitrary key/value pairs related to the
+  #   exception.
+  #
+  # - A transient flag, defaulting to true, indicating whether the exception
+  #   should be considered transient or final.  This can be used when rescuing
+  #   the exception to determine whether the operation should be retried.
+  #
+  # - An HTTP status, defaulting to 500.  This can be used to generate the
+  #   response payload if the exception is rescued as part of an HTTP request.
+  #
+  # - An optional static message, definable by subclasses, to be included
+  #   automatically in the details hash.  This removes the need to pass this
+  #   message in explicitly during initialization.
+
+  class Error < StandardError
+
+    # @return [Class] Class in which the exception occurred
+    attr_reader :klass
+    # @return [Hash] Arbitrary details about the exception
+    attr_reader :details
+    # @return [TrueClass, FalseClass] Whether the error is transient
+    attr_reader :transient
+    # @return [Integer] Corresponding HTTP status
+    attr_reader :http_status
```

Comment:
Should all errors have an http status? What about errors encountered outside the context of an http request?

Reply:
Not necessarily, but there may be classes tightly coupled to a Rails controller (e.g. parameter deserializer) that would want to specify a status.  Having this with a reasonable default allows code that cares about this to check, and code that doesn't not to.

Reply:
It does seem useful. What do you think about making the default be `nil` rather than `500` in recognition of the fact that only some errors are encountered in the context of an http request?

Reply:
Hmm.... How do you see the following situation being handled?

A low-level piece of code, far-removed from the top of the call stack, raises an exception.  It doesn't set an HTTP status code, because it doesn't have that context.  Eventually, the exception is rescued at the top of the stack, in a controller.  How does the controller decide what to render?  I suppose it could do something like

```rb
render json: { ... }, status: ex.status || 500
```

But then that's the same result as if the exception's default were 500.

This default value was also motivated by the fact that Rails controllers render unrescued exceptions with a 500 code.

Reply:
> Eventually, the exception is rescued at the top of the stack, in a controller. How does the controller decide what to render?

I'm not sure I'm following. How do you envision `YetiException::Error` being involved in this scenario where a low-level piece of code raises an exception? Are we rescuing the low-level error and then wrapping it in a `YetiException::Error` and re-raising?

Reply:
I'm assuming everything in the app is using `YetiException::Error`, whether it raises it explicitly or rescues/wraps/reraises.

The overall idea here is to make it so that no plumbing needs to be done between the source of the exception and where it's eventually handled.  This also allows the number of log messages to be reduced without losing any context about the error.  Here's a slightly-contrived example showing a lot of plumbing:

Before:
```rb
class Top
  class TopError < StandardError
    def initialize
      super('Error in Top')
    end
  end

  def self.call(thing)
    Mid.call(thing)
  rescue => ex
    log_error('Oops!', ex)
    raise TopError
  end
end

class Mid
  def self.call(thing)
    Bottom.call(thing)
  rescue => ex
    log_error('Boom!', ex)
    raise
  end
end

class Bottom
  class BottomError < StandardError; end

  def self.call(thing)
    if thing.bad?
      log_error({ msg: 'Bad thing', thing: thing })
      raise BottomError
    end
    ...
  end
end

def run(thing)
  Top.call(thing)
rescue => ex
  log_error('Bang!', ex)
end
```

After:
```rb
class Top
  def self.call(thing)
    Mid.call(thing)
  end
end

class Mid
  def self.call(thing)
    Bottom.call(thing)
  end
end

class Bottom
  class BottomError < YetiException::Error
    def msg
      'Bad thing'
    end
  end

  def self.call(thing)
    if thing.bad?
      raise_exception(BottomError, { thing: thing })
    end
    ...
  end
end

def run(thing)
  Top.call(thing)
rescue YetiException::Error => ex
  log_error(ex.details, ex)
end
```

Reply:
Thanks for that very detailed reply. That all makes sense to me.

In you example, would you envision the `Bottom` class choosing an http error code just in case the error was encountered while fulfilling an http request? I would imagine that `Bottom` wouldn't have any _other_ knowledge of http request or know who was invoking it.

What does it look like in the controller? Will `YetiException` also offer some way to mix in a convenient `rescue_from YetiException::Error` handler that is expected to be used by rails controller?

Reply:
I doubt in this case `Bottom` would choose a specific HTTP error code, unless `Bottom` was something very specialized like a parameter parser.

The readme updates in #3 include an example of using this in a Rails controller.

Reply:
```
  # For the YetiException exceptions with richer data, use the specified
  # details and status.
  rescue_from YetiException::Error do |ex|
    render json: { error: ex.details }, status: ex.status
  end
```

> I doubt in this case Bottom would choose a specific HTTP error code

So then in this example, `Bottom` is emitting an error with `http_status` 500, and the controller passes that on to the client directly.

My 2 cents: I think most yesware code that raises an exception won't have any knowledge of http connections and shouldn't be responsible for choosing http error codes. In some scenarios, the same exception might correctly result in different http error codes depending on the specifics of how it's used by a controller. Setting an http error code should be the concern of rails controllers.

That said, it doesn't really matter a whole lot in the grand scheme. I'm ok with this as-is if you feel strongly that this is the correct approach.

Reply:
I'm just saying this is no different that default Rails conventions, in which a 500 is effectively the default.

Reply:
I presume rails handles that in the controller, and not in the underlying exception class.

Reply:
I'm thinking of it this way: For a class to set an http error code in a friendly manner, it may [`require 'net/http'`](https://www.rubydoc.info/stdlib/net/Net) and then use one of the many friendly constants provided within, like `Net::HTTPUnauthorized` or `Net::HTTPInternalServerError`. That's considerably easier for future devs to read than using a magic number like `418`. 

But if we do that, then we've introduced `net/http` as a dependency to a wide swath of yesware code. Identity verification now emits `Net::HTTPUnauthorized` or `Net::HTTPForbidden`. Database context classes now emit `Net::HTTPNotFound` or `Net:: HTTPUnprocessableEntity`. It giving knowledge of http-related concerns to classes that would otherwise have no need for them.

Reply:
> It's giving knowledge of http-related concerns to classes that would otherwise have no need for them

I see it as "it's giving access of http-related concerns to all classes, many of which have no need for them".  But it's not forcing it upon them - there are no library dependencies, and the status has a default value.  I expect most usage of these exceptions will not set the value.

But I can see a place in controller code (either directly in the controller class or in classes tightly coupled to it) that would have enough knowledge of what error code to return - even if it's just 400 (i.e. default for `YetiException::ClientError`).  That can remove a lot of `render status: <whatever>` inline in controller methods, delegating all that to the global handler.

I think we're just going to have to agree to disagree on this one.